neuraxle/checkpoints.py

```diff
@@ -0,0 +1,148 @@
+"""
+Neuraxle's Checkpoint Classes
+====================================
+The checkpoint classes used by the checkpoint pipeline runner
+..
+    Copyright 2019, Neuraxio Inc.
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+        http://www.apache.org/licenses/LICENSE-2.0
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+"""
+from abc import abstractmethod
+import os
+import pickle
+from typing import Tuple, Any
+
+from neuraxle.pipeline import ResumableStep
+
+DEFAULT_CACHE_FOLDER = os.path.join(os.getcwd(), 'cache')
+
+
+class BaseCheckpointStep(ResumableStep):
+    """
+    Base class for a checkpoint step that can persists the received data inputs, and expected_outputs
+    to eventually be able to load them using the checkpoint pipeline runner.
+    """
+
+    def __init__(self, force_checkpoint_name: str = None):
+        ResumableStep.__init__(self)
+        self.force_checkpoint_name = force_checkpoint_name
+
+    def fit(self, data_inputs, expected_outputs=None) -> 'BaseCheckpointStep':
+        """
+        Save checkpoint for data inputs and expected outputs so that it can
+        be loaded by the checkpoint pipeline runner on the next pipeline run
+        :param expected_outputs: initial expected outputs of pipeline to load checkpoint from
+        :param data_inputs: data inputs to save
+        :return: self
+        """
+        self.save_checkpoint(data_inputs)
+
+        return self
+
+    def transform(self, data_inputs):
+        """
+        Save checkpoint for data inputs and expected outputs so that it can
+        be loaded by the checkpoint pipeline runner on the next pipeline run
+        :param data_inputs: data inputs to save
+        :return: data_inputs
+        """
+        self.save_checkpoint(data_inputs)
+
+        return data_inputs
+
+    def load_checkpoint(self, data_inputs) -> Tuple[list, Any]:
+        """
+        Load the checkpoint step data inputs.
+        There is no steps left to do when we are loading an actual checkpoint
+        instead of a pipeline that can contain checkpoints.
+        :param data_inputs:
+        :return: steps_left_to_do, data_inputs_checkpoint
+        """
+        return [], self.read_checkpoint()
```

Comment:
Could we avoid returning 2 items? Any other solution? I feel like load checkpoint should only be in the Pipeline class, and not inherited here. Possible breach of the LSP or ISP principles because of that.